src/components/admin/AdminUsersPanel.jsx

```diff
@@ -0,0 +1,203 @@
+import React, { useState, useEffect } from 'react'
+import PropTypes from 'prop-types'
+import ReactDOM from 'react-dom'
+import {
+  Card,
+  CardHeader,
+  CardTitle,
+  CardBody,
+  ListGroup,
+  ListGroupItem,
+  InputGroup,
+  InputGroupAddon,
+  Button,
+  FormText,
+} from 'reactstrap'
+import { useInput } from 'react-hanger'
+import { useDebounce } from 'use-debounce'
+import { CancelToken } from 'axios'
+import FlipMove from 'react-flip-move'
+import { AsyncTypeahead, Highlighter } from 'react-bootstrap-typeahead'
+
+import axios from '../../actions/axios'
+import Loading from '../Loading'
+
+import 'react-bootstrap-typeahead/css/Typeahead.css'
+import 'react-bootstrap-typeahead/css/Typeahead-bs4.css'
+import RemoveableUserItem from '../RemoveableUserItem'
+
+const AdminUsersPanel = props => {
+  const [admins, setAdmins] = useState([])
+  const [adminsLoading, setAdminsLoading] = useState(true)
+  const netidInput = useInput('')
+  const [netidQuery] = useDebounce(netidInput.value, 500)
+  const [userSuggestions, setUserSuggestions] = useState([])
+  const [userSuggestionsLoading, setUserSuggestionsLoading] = useState(false)
+  const [pendingAdmin, setPendingAdmin] = useState([])
+
+  useEffect(() => {
+    axios
+      .get('/api/users/admins')
+      .then(res => {
+        setAdmins(res.data)
+        setAdminsLoading(false)
+      })
+      .catch(err => {
+        console.error(err)
+      })
+  }, [])
+
+  useEffect(() => {
+    if (!netidQuery) {
+      return () => {}
+    }
+    const source = CancelToken.source()
+    setUserSuggestionsLoading(true)
+    axios
+      .get('/api/autocomplete/users', {
+        params: {
+          q: netidQuery,
+        },
+        cancelToken: source.token,
+      })
+      .then(res => {
+        ReactDOM.unstable_batchedUpdates(() => {
+          setUserSuggestions(
+            res.data.filter(user => {
+              return (
+                admins.findIndex(admin => admin.netid === user.netid) === -1
+              )
+            })
+          )
+          setUserSuggestionsLoading(false)
+        })
+      })
+      .catch(err => {
+        console.error(err)
+      })
+    return () => {
+      source.cancel()
+    }
+  }, [netidQuery])
+
+  const addAdmin = () => {
+    const user = pendingAdmin[0]
+    axios
+      .put(`/api/users/admins/${user.id}`)
+      .then(res => {
+        setPendingAdmin([])
+        setAdmins([...admins, res.data])
```

Comment:
Super small nit but after adding an admin, the typeahead refresh delay is large enough so that you can add the same user again, which renders weirdly because the server returns an empty response in that case. Maybe we could just validate the response data before updating the list of admins.

Reply:
Turns out the typeahead component can do filtering itself even when reading from its internal cache - I updated it to use that and am pretty sure that fixed things. Let me know if you can still reproduce this.